lib/Dialect/ESI/runtime/cpp/include/esi/Utils.h

```diff
@@ -17,12 +17,80 @@
 #define ESI_UTILS_H
 
 #include <cstdint>
+#include <functional>
+#include <mutex>
+#include <optional>
+#include <queue>
 #include <string>
 
 namespace esi {
 namespace utils {
 // Very basic base64 encoding.
 void encodeBase64(const void *data, size_t size, std::string &out);
+
+/// Thread safe queue. Just wraps std::queue protected with a lock. Long term,
+/// we need to avoid copying data. It has a lot of data copies currently.
+template <typename T>
+class TSQueue {
+  using Lock = std::lock_guard<std::mutex>;
+
+  /// The queue and its mutex.
+  mutable std::mutex qM;
```

Comment:
Why mutable?

Reply:
`empty()` needs to lock this and it's a const function.